### PR TITLE
[11.0] [MIG] sale_start_end_dates

### PR DESCRIPTION
**Pull request by murtuzasaleh**
Sale Start End Dates
====================

- This module adds the fields *start date* and *end date* on sale order
lines. Upon invoice creation, the value of the start/end dates of the
sale order line is copied to the start/end dates of invoice lines. This
module is a technical module: it is designed to be used with other
modules such as the *sale_rental* module.

Migration of module Sale Start End Dates v10 to v11.

### CODE REVIEW

File: sale_start_end_dates/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'Sale Start End Dates',
+    'version': '11.0.1.0.0',
+    'category': 'Sales Management',
+    'license': 'AGPL-3',
+    'summary': 'Adds start date and end date on sale order lines',
+    'author': 'Akretion, Sodexis, Odoo Community Association (OCA)',
+    'website': 'http://www.akretion.com',
```
Comment (JayVora-SerpentCS) [inline sale_start_end_dates/__manifest__.py]: https://github.com/OCA/sale-workflow

File: sale_start_end_dates/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+from dateutil.relativedelta import relativedelta
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    default_start_date = fields.Date(string='Default Start Date')
+    default_end_date = fields.Date(string='Default End Date')
+
+    @api.one
```
Comment (JayVora-SerpentCS) [inline sale_start_end_dates/models/sale.py]: api.multi

File: sale_start_end_dates/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        states={'draft': [('readonly', False)]})
+    end_date = fields.Date(
+        string='End Date', readonly=True,
+        states={'draft': [('readonly', False)]})
+    number_of_days = fields.Integer(string='Number of Days')
+    must_have_dates = fields.Boolean(
+        related='product_id.must_have_dates', readonly=True)
+
+    @api.one
+    @api.constrains('start_date', 'end_date', 'number_of_days')
```
Comment (JayVora-SerpentCS) [inline sale_start_end_dates/models/sale.py:54]: use multi

**Review by JayVora-SerpentCS — CHANGES_REQUESTED** [review]

**Review by JayVora-SerpentCS — APPROVED** [review]

File: sale_start_end_dates/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
+# Copyright 2014-2016 Akretion (http://www.akretion.com)
+# @author Alexis de Lattre <alexis.delattre@akretion.com>
+# Copyright 2016 Sodexis (http://sodexis.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, _
```
Comment (nikul-serpentcs) [inline sale_start_end_dates/models/sale.py]: @murtuzasaleh Use alphabetic naming convention.

File: sale_start_end_dates/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                    "the End Date (%s), but the number of days field "
+                    "has a value of %d days.")
+                    % (self.product_id.name, days_delta, self.start_date,
+                        self.end_date, self.number_of_days))
+
+    @api.multi
+    def _prepare_invoice_line(self, qty):
+        self.ensure_one()
+        res = super(SaleOrderLine, self)._prepare_invoice_line(qty)
+        if self.must_have_dates:
```
Comment (nikul-serpentcs) [inline sale_start_end_dates/models/sale.py]: @murtuzasaleh here check `if self.must_have_dates and self.start_date and self.end_date`

**Review by nikul-serpentcs — APPROVED** [review]
LGTM

**Review by YogeshMahera-SerpentCS — APPROVED** [review]
LGTM

**Pull request merged**